// @testomatio/reporter/lib/reporter.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.linkJira = exports.linkTest = exports.label = exports.step = exports.meta = exports.logger = exports.log = exports.artifact = exports.STATUS = exports.Client = void 0;
const client_js_1 = __importDefault(require("./client.js"));
exports.Client = client_js_1.default;
const TestomatioConstants = __importStar(require("./constants.js"));
const index_js_1 = require("./services/index.js");
const reporter_functions_js_1 = __importDefault(require("./reporter-functions.js"));
exports.STATUS = TestomatioConstants.STATUS;
exports.artifact = reporter_functions_js_1.default.artifact;
exports.log = reporter_functions_js_1.default.log;
exports.logger = index_js_1.services.logger;
exports.meta = reporter_functions_js_1.default.keyValue;
exports.step = reporter_functions_js_1.default.step;
exports.label = reporter_functions_js_1.default.label;
exports.linkTest = reporter_functions_js_1.default.linkTest;
exports.linkJira = reporter_functions_js_1.default.linkJira;
/**
 * @typedef {typeof import('./reporter-functions.js').default.artifact} ArtifactFunction
 * @typedef {typeof import('./reporter-functions.js').default.log} LogFunction
 * @typedef {typeof import('./services/index.js').services.logger} LoggerService
 * @typedef {typeof import('./reporter-functions.js').default.keyValue} MetaFunction
 * @typedef {typeof import('./reporter-functions.js').default.step} StepFunction
 * @typedef {typeof import('./reporter-functions.js').default.label} LabelFunction
 */
module.exports = {
    /**
     * @deprecated Use `log` or `testomat.log`
     */
    testomatioLogger: index_js_1.services.logger,
    artifact: reporter_functions_js_1.default.artifact,
    log: reporter_functions_js_1.default.log,
    logger: index_js_1.services.logger,
    meta: reporter_functions_js_1.default.keyValue,
    step: reporter_functions_js_1.default.step,
    label: reporter_functions_js_1.default.label,
    linkTest: reporter_functions_js_1.default.linkTest,
    linkJira: reporter_functions_js_1.default.linkJira,
    TestomatioClient: client_js_1.default,
    STATUS: exports.STATUS,
};
